src/transformers/utils/doctest_utils.py

```diff
@@ -0,0 +1,175 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+    Utils to run the documentation tests without having to overwrite any files.s
+
+    The doc precossing function can be run on a list of files and/org directories of files. It will recursively check
+    if the files have a python code snippet by looking for a ```python or ```py syntax. The script will add a new line
+    before every python code ending ``` line to make the docstrings ready for pytest doctests. However, we don't want
+    to have empty lines displayed in the official documentation which is why the new code is written to a temporary
+    directory / is tested on the fly depending on the configuration.
+
+    When debugging the doc tests locally, the script should automatically determine which files should be processed
+    based on the modified files. It should also run the tests on the fly and delete the temp file when finished.
+"""
+import doctest
+import inspect
+import os
+import re
+from typing import Iterable
+
+from _pytest.doctest import (
+    Module,
+    _get_checker,
+    _get_continue_on_failure,
+    _get_runner,
+    _is_mocked,
+    _patch_unwrap_mock_aware,
+    get_optionflags,
+    import_path,
+)
+from _pytest.outcomes import skip
+from pytest import DoctestItem
+
+
+def preprocess_string(string, skip_cuda_tests):
+    codeblock_pattern = r"(```(?:python|py)\s*\n\s*>>> )((?:.*?\n)*?.*?```)"
+    codeblocks = re.split(re.compile(codeblock_pattern, flags=re.MULTILINE | re.DOTALL), string)
+    for i, codeblock in enumerate(codeblocks):
+        if ">>>" in codeblock and "```py" in codeblock:
+            # let's add everything we need here, but the DATASET_VERBOSITY should work.
+            finale_block = codeblock
+            finale_block += "import transformers;transformers.logging.set_verbosity_error();"
+            finale_block += (
+                "import datasets;datasets.logging.set_verbosity_error();from contextlib import redirect_stdout;"
+            )
+            finale_block += "import huggingface_hub;huggingface_hub.logging.set_verbosity_error();"
+            codeblocks[i] = finale_block
+        if "load_dataset(" in codeblock:
+            codeblocks[i] = re.sub(
+                r">>> (.*)load_dataset\(", r">>> with redirect_stdout(None): \1load_dataset(", codeblock
+            )
+
+        if "cuda" in codeblock and ">>>" in codeblock and skip_cuda_tests:
+            if 'device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")' in codeblock:
+                continue
+            line_num = "".join(codeblocks[: i + 1]).count("\n")
+            codeblocks[
+                i - 1
+            ] += f'from pytest import skip;skip("Codeblock line {line_num} uses `cuda`. Skipping",allow_module_level=True);'
```

Comment:
A bit flaky as the GPU could also be used with a `to(0)`.

Reply:
Added check for device = 0 and `to(0)`.